### PR TITLE
Read new data format with multi hadrons in the final state

### PR DESCRIPTION
**Pull request by jaychan-hep**
Add the new dataset class `HerwigEventMultiHadronDataset` to read the new data format that can have more than 2 hadrons per cluster decay. The data module class is the same as what we used for the event GAN `EventGANDataModule`. We can create a new one and modify from there if needed.  The processed data are saved as the `pyg` format. Each pyg contains the following keys:

- `x`: conditional variables (each row is a cluster) [cluster_e, cluster_px, cluster_py, cluster_pz, q1_pid, q2_pid, q_phi, q_theta] (q_phi and q_theta are in the cluster rest frame)
- `had_kin_rest_frame`: hadron 4 momentum in the cluster rest frame (each row is a hadron)
- `had_kin`: hadron 4 momentum in the lab frame (each row is a hadron)
- `had_type_indices`: hadron type indices (the hadron type label converted from pid with pids_to_ix mapping)
- `cluster_labels`: to help match hadrons to cluster. The hadrons with the same cluster label and the sane event label (using the keyword `batch`) belong to the same cluster.

The yaml files `Herwig_event_multihadron.yaml` is the example config for this new data module class.
Also uploaded an example data format in `data/Herwig/raw/AllClusters_paper4.dat`

Something to note:
- There is no GAN model to train with this new data yet. We need to create a new GAN model to train with multi hadrons.
- Currently the processed data will save all hadron types (as opposed to paper 3 where we only selected the top K hadron types). Need to modify it further if we want to select only the top K hadron types

Also add the notebook `test_data.ipynb` to test and plot the input sample

### CODE REVIEW

File: configs/datamodule/herwig_event_multihadron.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+batch_size: 1
+num_workers: 1
+pin_memory: False
+train_val_test_split: [5, 2, 3]
+frac_data_used:
+cond_dataset:
+  _target_: hadml.datamodules.components.herwig.HerwigEventMultiHadronDataset
+  root: ${paths.data_dir}Herwig/
+  raw_file_list:
+    # - "ClusterTo2Pi0_new.dat"
```
Comment (xju2) [inline configs/datamodule/herwig_event_multihadron.yaml:11]: remove un-used code.

File: configs/datamodule/herwig_event_multihadron.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    - "AllClusters_paper4.dat"
+  processed_file_name: "herwig_multihadron_graph_cond_data.pt"
+
+obs_dataset:
+  _target_: hadml.datamodules.components.herwig.HerwigEventMultiHadronDataset
+  root: ${paths.data_dir}Herwig/
+  raw_file_list:
+    # - "ClusterTo2Pi0_S112*.dat"
```
Comment (xju2) [inline configs/datamodule/herwig_event_multihadron.yaml:19]: remove it.

**Review by xju2 — APPROVED** [review]
Looks good. Some minor comments.

File: configs/datamodule/herwig_event_multihadron.yaml (hunk 1 above)
Comment (jaychan-hep) [reply configs/datamodule/herwig_event_multihadron.yaml]: Done

File: configs/datamodule/herwig_event_multihadron.yaml (hunk 2 above)
Comment (jaychan-hep) [reply configs/datamodule/herwig_event_multihadron.yaml]: Done

**Comment by jaychan-hep** [comment]
Thanks @xju2 I have removed those unused lines and merged it now.

**Pull request merged**